File: 2023/golang/day1/day1.go
package day1

import (
	"regexp"
	"strconv"
	"strings"
)

func Solve(input string, partTwo bool) int {
	lines := strings.Split(input, "\n")
	var numbers []int
	for _, line := range lines {
		var lineNumbers string
		for i := 0; i < len(line); i++ {
			number, ok := extractNumbers(line[i:], partTwo)
			if ok {
				lineNumbers += number
			}
		}
		if len(lineNumbers) > 0 {
			calibrationNumber, err := strconv.Atoi(lineNumbers[0:1] + lineNumbers[len(lineNumbers)-1:])
			if err != nil {
				continue
			}
			numbers = append(numbers, calibrationNumber)
		}
	}

	var sum int
	for _, num := range numbers {
		sum += num
	}

	return sum
}

func extractNumbers(line string, partTwo bool) (string, bool) {
	re := regexp.MustCompile(`^\d`)
	switch {
	case re.MatchString(line):
		return line[0:1], true
	case strings.HasPrefix(line, "one") && partTwo:
		return "1", true
	case strings.HasPrefix(line, "two") && partTwo:
		return "2", true
	case strings.HasPrefix(line, "three") && partTwo:
		return "3", true
	case strings.HasPrefix(line, "four") && partTwo:
		return "4", true
	case strings.HasPrefix(line, "five") && partTwo:
		return "5", true
	case strings.HasPrefix(line, "six") && partTwo:
		return "6", true
	case strings.HasPrefix(line, "seven") && partTwo:
		return "7", true
	case strings.HasPrefix(line, "eight") && partTwo:
		return "8", true
	case strings.HasPrefix(line, "nine") && partTwo:
		return "9", true
	default:
		return "0", false
	}
}

File: 2023/typescript/src/challenges/day1.ts
import { Challenge } from './challenge';

export class Day1 extends Challenge {
  private partTwo: boolean;
  private lines: Array<string>;

  constructor(input: string, partTwo: boolean) {
    super();
    this.lines = input.split('\n');
    this.partTwo = partTwo;
  }

  solve = (): number => {
    return this.lines
      .map((line) => {
        const numbers = this.extractNumbers(line);
        return parseInt(numbers.slice(0, 1) + numbers.slice(-1));
      })
      .reduce((sum, num) => sum + num, 0);
  };

  private extractNumbers = (line: string): string => {
    let tempLine = line;

    if (this.partTwo) {
      for (let i = 0; i < tempLine.length; i++) {
        let num: string = '';
        switch (true) {
          case tempLine.slice(i).startsWith('one'):
            num = '1';
            break;
          case tempLine.slice(i).startsWith('two'):
            num = '2';
            break;
          case tempLine.slice(i).startsWith('three'):
            num = '3';
            break;
          case tempLine.slice(i).startsWith('four'):
            num = '4';
            break;
          case tempLine.slice(i).startsWith('five'):
            num = '5';
            break;
          case tempLine.slice(i).startsWith('six'):
            num = '6';
            break;
          case tempLine.slice(i).startsWith('seven'):
            num = '7';
            break;
          case tempLine.slice(i).startsWith('eight'):
            num = '8';
            break;
          case tempLine.slice(i).startsWith('nine'):
            num = '9';
            break;
        }
        if (!num) continue;

        const temp = tempLine.split('');
        temp[i] = num;
        tempLine = temp.join('');
      }
    }
    return tempLine.replace(/\D/g, '');
  };
}
